### PR TITLE
tooling for merging multiple ONNX models, greedy search, simpler creation of new ONNX models

### PR DESCRIPTION
**Pull request by frankseide**
This PR is a collaboration with equal contributions from Wenbing Li (wenbingl) and myself.

This PR adds tooling for creating new ONNX models by combining existing ones, and also simplifies creation of new models from scratch. This was originally aimed at creating a single ONNX model that implements greedy search by combining three sub-models exported from the Marian-NMT toolkit.

This PR adds a new helper library, `onnx_fx.py`, which implements two main helper classes for creating and merging models, `Tensor`, and `Graph`.

`Graph` wraps ONNX models. It allows to load an existing ONNX model, or to create a new ONNX model via tracing by invoking a Python function, and save it. The latter is modeled after CNTK's `Function` decorator. Example:
```
    @Graph.trace(outputs="s")
    def f(x, y):
       return x + y
```
This creates a `Graph` instance called `f` which contains an ONNX graph with two inputs `x` and `y` that adds the two together.

A `Graph` instance can be invoked like a function, e.g. `x.ox.abs(f(x, y) + 1.0)`. If this is done while tracing, the graph will get inlined into the outer graph. If the invocation uses actual numbers (Numpy arrays), then it will get evaluated via ORT, e.g. `f(3.0, 5.0)` will return 8.0, which is useful for testing.

`Tensor` is a shallow wrapper for all tensor arguments and return values for all ONNX operators. it provides overloads for the common infix operators, and thus allows to write `x + y` instead of `ox.add([x, y])`. It also provides chaining overloads, allowing to say `logits.argmax()` instead of having to write `ox.argmax(logits)`. Lastly, it implements automatic type casts from numerals, so that one can write `x + 1` instead of `x + ox.constant(np.array(1), dtype=ox.int64)`.

To combine multiple ONNX models into one, one loads those models into `Graph` instances, and then traces a new `Graph` that invokes the models in the desired way.

As an example, this was used to load three models exported from the Marian Neural Machine Translation (NMT) toolkit, one representing an NMT encoder, and two representing parts of a decoding step. These were bound together into a single ONNX model that implements greedy search, by first invoking the encoder, and then invoking the decoding steps from inside an ONNX `Loop` construct.

### CODE REVIEW

**Comment by wenbingl** [comment]
Can you discard the changes on these files? 
onnxconverter_common/optimizer.py 
onnxconverter_common/topology.py 
temp_onnx_fx_test.py 

The change on these files are not complete yet and it breaks the CI tests. I'll add some tests to enable the change step by step in the future.
